core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs

```diff
@@ -0,0 +1,660 @@
+            .try_iter()
+            .map(|work| work.ids.len())
+            .collect();
+        assert_eq!(thread0_work_counts, [TARGET_NUM_TRANSACTIONS_PER_BATCH; 4]);
+    }
+
+    #[test]
+    fn test_schedule_simple_thread_selection() {
+        let (mut scheduler, work_receivers, _finished_work_sender) = create_test_frame(2);
+        let mut container =
+            create_container((0..4).map(|i| (Keypair::new(), [Pubkey::new_unique()], 1, i)));
+
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 4);
+        assert_eq!(collect_work(&work_receivers[0]).1, [txids!([3, 1])]);
+        assert_eq!(collect_work(&work_receivers[1]).1, [txids!([2, 0])]);
+    }
+
+    #[test]
+    fn test_schedule_look_ahead() {
+        let (mut scheduler, work_receivers, _finished_work_sender) = create_test_frame(2);
+
+        let accounts = (0..4).map(|_| Keypair::new()).collect_vec();
+        let mut container = create_container([
+            (&accounts[0], &[accounts[1].pubkey()], 1, 2),
+            (&accounts[2], &[accounts[3].pubkey()], 1, 1),
+            (&accounts[1], &[accounts[2].pubkey()], 1, 0),
+        ]);
+
+        // high priority transactions [0, 1] do not conflict, and would be
+        // scheduled to *different* threads without chain-id look-ahead.
+        // Because low priority transaction [2] conflicts with both, it will
+        // cause transaction [1] to be scheduled onto the same thread as
+        // transaction [0].
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 3);
+        assert_eq!(
+            collect_work(&work_receivers[0]).1,
+            [txids!([0, 1]), txids!([2])]
+        );
+    }
+
+    #[test]
+    fn test_schedule_priority_guard() {
+        let (mut scheduler, work_receivers, finished_work_sender) = create_test_frame(2);
+        // intentionally shorten the look-ahead window to cause unschedulable conflicts
+        scheduler.look_ahead_window_size = 2;
+
+        let accounts = (0..8).map(|_| Keypair::new()).collect_vec();
+        let mut container = create_container([
+            (&accounts[0], &[accounts[1].pubkey()], 1, 6),
+            (&accounts[2], &[accounts[3].pubkey()], 1, 5),
+            (&accounts[4], &[accounts[5].pubkey()], 1, 4),
+            (&accounts[6], &[accounts[7].pubkey()], 1, 3),
+            (&accounts[1], &[accounts[2].pubkey()], 1, 2),
+            (&accounts[2], &[accounts[3].pubkey()], 1, 1),
+        ]);
+
+        // high priority transactions [0, 1, 2, 3] do not conflict, and are
+        // scheduled onto threads in a round-robin fashion.
+        // The look-ahead window is intentionally shortened, which leads to
+        // transaction [4] being unschedulable due to conflicts with [0] and [1],
+        // which were scheduled to different threads.
+        // Transaction [5] is technically schedulable, but will not be scheduled
```

Comment:
🤔 I think I copied my comment from the wrong test-case in my messed up branch.
I actually do want a test where [5] would be schedulable, but isn't because of priority-ordering...I'll adjust this case.